src/coreclr/jit/hwintrinsiccodegenxarch.cpp

```diff
@@ -2125,38 +2127,116 @@ void CodeGen::genFMAIntrinsic(GenTreeHWIntrinsic* node)
     regNumber op1Reg;
     regNumber op2Reg;
 
+    regNumber op1NodeReg = op1->GetRegNum();
+    regNumber op2NodeReg = op2->GetRegNum();
+    regNumber op3NodeReg = op3->GetRegNum();
+
+    // For special case that targetReg equals to more than one op Regs
+    bool uniqueOverwrittenReg = true;
+    if ((op1NodeReg == op2NodeReg && targetReg == op1NodeReg) ||
+        (op1NodeReg == op3NodeReg && targetReg == op1NodeReg) || (op2NodeReg == op3NodeReg && targetReg == op2NodeReg))
+        uniqueOverwrittenReg = false;
+
     bool       isCommutative   = false;
     const bool copiesUpperBits = HWIntrinsicInfo::CopiesUpperBits(intrinsicId);
 
     // Intrinsics with CopyUpperBits semantics cannot have op1 be contained
     assert(!copiesUpperBits || !op1->isContained());
 
-    if (op2->isContained() || op2->isUsedFromSpillTemp())
+    // Intrinsics with CopyUpperBits semantics must have op1 as target
+    if ((targetReg == op1NodeReg || copiesUpperBits) && uniqueOverwrittenReg)
```

Comment:
I wonder if we could simplify some of this logic in the same way I described in https://github.com/dotnet/runtime/pull/58196/files#r750443089

That is, we basically have:
* `op1 = lhsMul`, `op2 = rhsMul`, `op3 = addend`

We need to pick an instruction form and swap the operands around to match that:
* `vfmadd132` - `emitOp1 = lhsMul | rhsMul`, `emitOp2 = addend`, `emitOp3 = lhsMul | rhsMul`
* `vfmadd213` - `emitOp1 = lhsMul | rhsMul`, `emitOp2 = lhsMul | rhsMul`, `emitOp3 = addend`
* `vfmadd231` - `emitOp1 = addend`, `emitOp2 = lhsMul | rhsMul`, `emitOp3 = lhsMul | rhsMul`

Where for `vfmadd132` and `vfmadd231`, the contained operand if any should be `emitOp3`.